File: Leetcode-one-by-one/src/Check If a String Contains All Binary Codes of Size K.java
class Solution {
    public boolean hasAllCodes(String s, int k) {
        int need = 1 << k;                 // total binary codes = 2^k
        boolean[] seen = new boolean[need];
        int mask = need - 1;               // to keep last k bits
        int hash = 0;
        int count = 0;

        for (int i = 0; i < s.length(); i++) {
            // build rolling hash
            hash = ((hash << 1) & mask) | (s.charAt(i) - '0');

            // start checking once window size reaches k
            if (i >= k - 1) {
                if (!seen[hash]) {
                    seen[hash] = true;
                    count++;
                    if (count == need) return true;
                }
            }
        }
        return false;
    }
}

File: Leetcode-one-by-one/src/Check If a String Contains All Binary Codes of Size K.swift
class Solution {
    func hasAllCodes(_ s: String, _ k: Int) -> Bool {
        let n = s.count
        if n < k { return false }

        let need = 1 << k                 // total binary codes = 2^k
        var seen = Array(repeating: false, count: need)
        let mask = need - 1               // keeps last k bits

        var hash = 0
        var count = 0

        let chars = Array(s)

        for i in 0..<n {
            // rolling hash using bitwise ops
            hash = ((hash << 1) & mask) | (chars[i] == "1" ? 1 : 0)

            if i >= k - 1 {
                if !seen[hash] {
                    seen[hash] = true
                    count += 1
                    if count == need {
                        return true
                    }
                }
            }
        }
        return false
    }
}
